### ISSUE TITLE
Non-responsive on yes

### ISSUE THREAD

**Issue by ony** [wez/wezterm#65]
## Did something not work the way you expected?

When running Linux command `yes` nothing is rendered and no way to terminate until it completes. Though time for processing is almost 2 times faster than in `alacritty`.

Try:

```
time yes | head -n 1000000
```

### What did you expect to happen?

See flow of `y` and ability to terminate before `head` terminates.

### What actually happened?

Nothing. Window doesn't repaint itself and doesn't react to `Ctrl-C` that usually manage to stop current command.  
Window updated and gains control only after command completes.

### Assumption

Looks like rendering and signals sending have lower priority than servicing PTY (or whatever is there).

**Comment by wez** [comment]
We have some options that are intended to help with this; look for the `ratelimit_output_bytes_per_second` option documented in this section:

https://wezfurlong.org/wezterm/configuration.html#misc-configuration

Would you mind experimenting a bit with that and see if that helps?  I'm not going to have a chance to investigate until the weekend!

**Comment by ony** [comment]
I'll try in ~10 hours. But having fixed rate limit doesn't sound good either. I.e. if your system able to serve output faster you should benefit from that. Though it still make sense to limit redrawing to the frame limit of display or perception limits of human sometimes.

**Comment by ony** [comment]
It looks like `ratelimit_output_bytes_per_second` just limits how much characters it can output in total.

After some time it draws output and hangs (at least for a minute). Re-paint, scrolling, and selection works.  
I also confirm that Ctrl-C results in command being interrupted (disappearing from running processes). And when I type something blindly in terminal it actually executes commands but no update in window content.

**Comment by wez** [comment]
I think in your initial scenario wezterm had already read a lot of data from the output stream prior to the ctrl-c being registered, so it still had to parse and display the result; a terminal doesn't know whether it is OK to do something like discard data that it has read in the event of an interrupt (and actually doesn't know that it is generating an interrupt: that is something that the kernel is interpreting and propagating), so if we read a few MB of data already we still need to process it.

With your tests with much smaller rate limits, I'm not super sure what to make of your findings yet and will re-read and try similar tests locally when I have a bit of time for that.

Thanks for looking into this!

**Comment by ony** [comment]
I checked. Again by running `yes` and after pressing Enter I don't even see how cursor moves to the next line to feed shell with a command. In separate console I checked `pgrep -a yes` and it was still there after pressing Ctrl-C on wezterminal. So it definitely ignores user input and re-draw events from X11 and have to be killed.

I also checked if wezterm consumes more memory and it is stable at 100-200 MB (2 processes each).

I would say that issue is that terminal is not blocking program output by prioritizing reading from output over interacting with user.

This just to ensure that this is still considered as an issue :)

- If terminal unable to render output at the rate program produces it should throttle so it will keep up. Not true for wezterm.
- User interaction should have higher priority over everything since it is interactive application. Not true for wezterm.

**Comment by wez** [comment]
I've pushed a fix for a panic that was blocking all output processing when the data was too large to ever be admitted to the rate limiter.

I do also observe that the output parsing is starving processing other items in the spawn queue and this is particularly acute with the X11 front end; there's still more work to be done to fully address this.

**Comment by ony** [comment]
Works for me now. Default rate limit (I guess 2MByte/s) on my system takes 8% of user time and gives enough responsiveness.

P.S. Specifying very high limit to find fastest behavior gives 4.5 seconds vs 3.2 s for rxvt-unicode vs 3.1 s for alacritty.  
I need to tweak it a bit and I guess I got my rxvt-unicode replacement with proper emojis support and maybe I'll use ligatures. Thank you for this project.

**Comment by wez** [comment]
Thanks for reporting this issue and helping work through it!
There's probably room for improving the throughput in this scenario, but that's lower priority from my perspective than remaining responsive.
